### DISCUSSION TITLE
How can I use variables/dataframe generated in streamlit app to be used by Quarto in generating html report in streamlit ?

### DISCUSSION THREAD

**Discussion by johnsnow09** [quarto-dev/quarto-cli#11577] [Q&A]
### Description

I have a python streamlit webapp [link](https://healthquant.streamlit.app/Data_Analysis)  running and I want to add a feature to it to allow users to **download** `html` reports from it. The **downloadable** `html` should be able to use `variables, dataframe, plotly charts` created in **streamlit** using **.py** files running in the vm. Is it possible to use **Quarto** for this and how can I use those variables from streamlit to Quarto. Will there be any thing extra I need to do to use those variables, dataframe by name. 
For example if there is pandas dataframe `st.session_state.data_final` (st represents streamlit, session_state is to pass variables between pages in session and data_final is the dataframe name) will this be directly loaded/read/utilized within Quarto by mentioning its name  `st.session_state.data_final` ?  

Is there any blog or demo or document available related to this ?

**Comment by mcanouil** [comment]
To be sure I understood what you want to do:
You want to deploy a Streamlit app that can run Quarto CLI using some values/parameters coming straight from your app. This means your app is running on a server that has Quarto CLI installed and all possible dependencies (Jupyter engine?). Then your app needs to be able to execute Quarto CLI and retrieve the rendered file.

I'm afraid, you need to develop this yourself and it is unlikely you'll find anything about this specifically, but you might find things online about the key principle to execute OS command from the app (likely via `subprocess`)

```python
process = subprocess.Popen(["quarto", "..."], ...)
```

**Reply by johnsnow09** [reply]
Thank you @mcanouil I am able to do this on local machine as of now and hopefully will be able to do it on cloud as well. 
And I am already using `subprocess` for this but I am **not able to extract the variables/parameters value** within **quarto** file. Below is something that I have already used:

`streamlit py file` that creates **variable** and call **subrpocess** to execute **quarto render**
```
import streamlit as st
import subprocess

# creating test variable to store some value
st.session_state.variable_check = 7+3

# to print the value of variable in app
st.write(f"variable_check value is {st.session_state.variable_check}")

# terminal command required to execute quarto and create html file
cmd_str = "quarto render test-1.qmd --output test2_cmd_out.html"

subprocess.call(cmd_str)
``` 

`test-1.qmd` Quarto file that runs and generate output without calling external variables:
````
---
title: "Test File"
format: html
execute:
  echo: False
jupyter: python3
---

# Title

```{python}
calculation = 3+2
calculation
```
evaluate variable value out of code block (None of them is able to extract value):

`{python} calculation`
{calculation}
{{calculation}}
{{ <calculation> }}

````
Above quarto file is able to run and generates output as `test2_cmd_out.html` file

So how can I use the **variables/ parameters** in below file that originated in **streamlit.py** file. I have tried below **qmd** code but that **doesn't work**
````
---
title: "Test File"
format: html
execute:
  echo: False
jupyter: python3
---

# Title
Testing variable value below within python code block

```{python}
st.session_state.variable_check
```

Testing variable value below
st.session_state.variable_check
````

I think [this link](https://forum.posit.co/t/how-to-access-r-sessions-variables-from-inside-quarto-file/163019/2) refers to similar situation in R session with Quarto.

Please suggest on how to **call or evaluate** `variables/parameters` inside `.qmd` file that already exist in session. Thanks !!!

**Reply by mcanouil** [reply]
You might want to read the documentation of Quarto before using it.

- https://quarto.org/docs/computations/parameters.html
- shortcodes meta/var
- include

**Reply by johnsnow09** [reply]
I have gone through the documentation and I am not sure how the interaction between two different projects/frameworks streamlit and Quatro can be made possible in same parent project but I have tried by creating `_variables.yml` file with variable that was created from streamlit.

`streamlit` file to create variable
```
import streamlit as st
import subprocess

# creating test variable to store some value
st.session_state.variable_check = 7+3

# to print the value of variable in app
st.write(f"variable_check value is {st.session_state.variable_check}")

# terminal command required to execute quarto and create html file
cmd_str = "quarto render test-4.qmd --output test4_cmd_out.html"

subprocess.call(cmd_str)
```

`_variables.yml` file to bridge variable of two different:
```
variable_check: st.session_state.variable_check
```

`test-4.qmd file` to use variable from _variables.ym
```
---
title: "Test File"
format: html
execute:
  echo: False
jupyter: python3
---

# Title


this is the variable value coming from streamlit: {{< var variable_check >}}

```
 
this runs the file and create .html file but it doesn't give the variable value instead I got `this is the variable value coming from streamlit: ?var:variable_check`

Please suggest if its even possible to do what I am trying to.

Thanks !!!

**Reply by mcanouil** [reply]
Use additional CLI flags with whatever elements you need.

```sh
quarto render --help
```

Unfortunately, you did not share anything concrete thus no concrete answer.

Quarto has many options either from YAML or command line flags.
That's up to you to choose what are the most appropriate for your goal.
You already used `--output`, use others to achieve your goal.

**Reply by johnsnow09** [reply]
Thanks for sharing the command, I will check that. I am new to development field so please excuse my ignorance.

I thought I shared the scripts code. "With nothing concrete" are you referring to `Github` repo ?

If that's so then here is the Github link to access the sample scripts: [Github Repo](https://github.com/johnsnow09/quarto_streamlit_test)

**Reply by johnsnow09** [reply]
I understand I can send `parameter` value to the qmd doc along with output flag `quarto render test-6.qmd -P dummy_var:5 --output test6_cmd_out.html` but how can I send `dataframe` object created in `streamlit` to `qmd`. I need to send `dataframe` object to qmd **without saving it in directory** due to session constrains and If I am able to do this then it is all sorted out for me otherwise I will have stop giving any more time to Quarto and move to alternatives. 

Thanks for the help !!!

**Reply by mcanouil** [reply]
You need to figure out a way to represent the data frame in another way that works in Quarto CLI flags (YAML,json,base64,etc).

**Reply by johnsnow09** [reply]
Ok, converting to `json` or `YAML` shouldn't be a problem. 

1. 
I am using `polars` dataframe so I can do 
`df_json = df.write_json()` or  `df_yaml = yaml.dump(df.to_dict(orient='records'),default_flow_style=None)`

So now can I just pass json and yaml file **object** with it as shown in below commands or do i need to save them as files first? Because saving as file is an issue for me so I cant do that. 

`quarto render test-6.qmd --execute-param df_json --output test6_cmd_out.html`
or 
`quarto render test-6.qmd --execute-param df_yaml --output test6_cmd_out.html`

-----------------------------------------------------------------

2. If 1st step is done then can I just use below line of code to access the json file contents in qmd ?
 
```{python}
df_json
```

But Getting error: **ERROR: YAML file df_json not found.**

If this is not correct then can you show me the correct way. Would really Appreciate your help !!

**Reply by mcanouil** [reply]
Again that's up to you to implement what you want as you want.

Parameters can be pass to Quarto as file or string from the command line.

Note that you are now aware of what Quarto itself can do with the flags or YAML frontmatter in a document.

If you use Python, you get access to everything that Python allows you to do.

Now your question has nothing to do with Quarto and is more about coding what you want.
I'm afraid I don't have the time to think and code any solution for your very specific problem which is not really a Quarto problem.

**Reply by johnsnow09** [reply]
Sure,thanks !!

First in my earlier post I was told about Anaconda problems when I was facing `Quarto CLI installation issues` through `pip` but was **not informed about installation through Conda forge** . I **deleted Anaconda** after that and then later got to know it could have also been installed through **Conda forge** which I eventually did. Deleting Anaconda messed up my system and environments which I am still fixing after days.

Now when I was expecting some support in getting to see some example or 1 liner of code on how it can actually be implemented unfortunately it was not there. Even in below post for R exactly same issue was raised and 1 liner code was provided to show how it can be done as a demo but not in my case.
https://forum.posit.co/t/how-to-access-r-sessions-variables-from-inside-quarto-file/163019/2

If that's how the community support is provided where I end up wasting days then I think its better to stop at the beginning of its use where learning curve is at peak but point of no support means to better look for alternatives with more community support.

Thanks for all the replies !!!
